src/astarte_storage.c

```diff
@@ -0,0 +1,465 @@
+/*
+ * (C) Copyright 2023, SECO Mind Srl
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR Apache-2.0
+ */
+
+#include "astarte_storage.h"
+
+#include <esp_log.h>
+#include <nvs.h>
+#include <stdlib.h>
+#include <string.h>
+
+/************************************************
+ *        Defines, constants and typedef        *
+ ***********************************************/
+
+#define TAG "ASTARTE_STORAGE"
+
+#define NVS_NAMESPACE_BLOB "ASTARTE_NS_BLOB"
+#define NVS_NAMESPACE_I32 "ASTARTE_NS_I32"
+
+/************************************************
+ *         Global functions definitions         *
+ ***********************************************/
+
+// NOLINTNEXTLINE(misc-unused-parameters)
+astarte_err_t astarte_storage_open(astarte_storage_handle_t *handle)
+{
+#ifdef CONFIG_ASTARTE_USE_PROPERTY_PERSISTENCY
+    esp_err_t esp_err
+        = nvs_open_from_partition(CONFIG_ASTARTE_PROPERTY_PERSISTENCY_NVS_PARTITION_LABEL,
+            NVS_NAMESPACE_BLOB, NVS_READWRITE, &handle->nvs_handle_blob);
+    if (esp_err != ESP_OK) {
+        return ASTARTE_ERR;
+    }
+    esp_err = nvs_open_from_partition(CONFIG_ASTARTE_PROPERTY_PERSISTENCY_NVS_PARTITION_LABEL,
+        NVS_NAMESPACE_I32, NVS_READWRITE, &handle->nvs_handle_i32);
+    if (esp_err != ESP_OK) {
+        nvs_close(handle->nvs_handle_blob);
+        return ASTARTE_ERR;
+    }
+#endif
+    return ASTARTE_OK;
+}
+
+void astarte_storage_close(astarte_storage_handle_t handle)
+{
+    nvs_close(handle.nvs_handle_blob);
+    nvs_close(handle.nvs_handle_i32);
+}
+
+astarte_err_t astarte_storage_store_property(astarte_storage_handle_t handle,
+    const char *interface_name, const char *path, int32_t major, const void *data, size_t data_len)
+{
+    // Get the full key interface_name + path
```

Comment:
Why didn't you make the key with` major version` as well? It would simplify the code, removing the i32_storage_handle, duplicate entries etc.

Reply:
It would lead to a larger complexity for operations such as 'store'. This is due to the underlying implementation of `astarte_nvs_key_value.c`.
I made changes to store the major version in the binary blob together with the property data.